storage/database/db_manager.go

```diff
@@ -1885,20 +1843,44 @@ func (dbm *databaseManager) ReadPreimageFromOld(hash common.Hash) []byte {
 	return data
 }
 
+func (dbm *databaseManager) WriteTrieNode(hash common.Hash, node []byte) error {
+	dbm.lockInMigration.RLock()
+	defer dbm.lockInMigration.RUnlock()
+
+	if dbm.inMigration {
+		if err := dbm.getDatabase(StateTrieMigrationDB).Put(TrieNodeKey(hash), node); err != nil {
+			return err
+		}
+	}
+	return dbm.getDatabase(StateTrieDB).Put(TrieNodeKey(hash), node)
+}
+
+func (dbm *databaseManager) PutTrieNodeToBatch(batch Batch, hash common.Hash, node []byte) error {
+	return batch.Put(TrieNodeKey(hash), node)
+}
+
 // WritePreimages writes the provided set of preimages to the database. `number` is the
 // current block number, and is used for debug messages only.
-func (dbm *databaseManager) WritePreimages(number uint64, preimages map[common.Hash][]byte) {
+func (dbm *databaseManager) WritePreimages(number uint64, preimages map[common.Hash][]byte) error {
 	batch := dbm.NewBatch(StateTrieDB)
 	for hash, preimage := range preimages {
 		if err := batch.Put(preimageKey(hash), preimage); err != nil {
-			logger.Crit("Failed to store trie preimage", "err", err)
+			return err
```

Comment:
Any reason returning error here? Doesn't it have to handle the error in caller?

Reply:
Why return an error?
- There were two use cases
  - call WritePreimage() and not check error because logger.Crit() inside
  - manually use Batcher.Put and check error.
- I intended to keep only the second case because returning an error is better than silently dying with Crit().

Error handling
- There are total 3 call sites of WritePreimages. I made sure the callers check the error.
- At blockchain/blockchain.go:1556, toss up the error
- At blockchain/blockchain.go:1657, toss up the error
- At storage/statedb/database.go:818, writeBatchPreimages() has been returning the error before this PR.

Reply:
Oh, I missed the error handling logic in `blockchain.go`. Thanks.